base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1212,114 +1212,74 @@ fn fetch_blocks<T: BlockchainBackend>(
-    //     let (utxo_hashes, _) = utxo_cp.into_parts();
-    //     let utxos = fetch_outputs(db, utxo_hashes)?;
-    //     for utxo in utxos.0 {
-    //         if utxo.commitment == commitment {
-    //             return Ok(None);
-    //         }
-    //     }
-    //     for comm in utxos.1 {
-    //         if comm == commitment {
-    //             return Ok(None);
-    //         }
-    //     }
-    // }
-    // Err(ChainStorageError::ValueNotFound {
-    //     entity: "Utxo".to_string(),
-    //     field: "Commitment".to_string(),
-    //     value: commitment.to_hex(),
-    // })
+    match db.fetch_output(&commitment.to_vec()) {
+        Ok(output) => match output {
+            Some((_output, leaf)) => {
+                let header = db.fetch_header_containing_utxo_mmr(leaf as u64)?;
+                fetch_block_by_hash(db, header.hash().to_owned())
+            },
+            None => Ok(None),
+        },
+        Err(_) => Err(ChainStorageError::ValueNotFound {
+            entity: "Output".to_string(),
+            field: "Commitment".to_string(),
+            value: commitment.to_hex(),
+        }),
+    }
 }
 
 fn fetch_block_with_stxo<T: BlockchainBackend>(
-    _db: &T,
-    _commitment: Commitment,
+    db: &T,
+    commitment: Commitment,
 ) -> Result<Option<HistoricalBlock>, ChainStorageError>
 {
-    unimplemented!()
-    // let metadata = db.fetch_chain_metadata()?;
-    // let db_height = metadata.height_of_longest_chain.unwrap_or(0);
-    // let horizon_height = metadata.horizon_block(db_height);
-    // for i in (horizon_height..db_height).rev() {
-    //     let utxo_cp = fetch_checkpoint(db, MmrTree::Utxo, i)?;
-    //     let (_, deleted) = utxo_cp.into_parts();
-    //     let inputs = fetch_inputs(db, deleted)?;
-    //     for input in inputs {
-    //         if input.commitment == commitment {
-    //             return Ok(Some(fetch_block(db, i)?));
-    //         }
-    //     }
-    // }
-    // // data is not in the pruning horizon, we cannot check stxo's behind pruning horizon
-    // Err(ChainStorageError::ValueNotFound {
-    //     entity: "Utxo".to_string(),
-    //     field: "Commitment".to_string(),
-    //     value: commitment.to_hex(),
-    // })
+    let tip = db.fetch_chain_metadata()?.best_block().clone();
+    let block_data = db
+        .fetch_block_accumulated_data(&tip)?
+        .ok_or_else(|| ChainStorageError::ValueNotFound {
+            entity: "BlockAccumulatedData".to_string(),
+            field: "header_hash".to_string(),
+            value: tip.to_hex(),
+        })?;
+    match db.fetch_output(&commitment.to_vec()) {
+        Ok(output) => match output {
+            Some((_output, leaf)) => {
+                // check if output is spent as of latest block
+                if !block_data.deleted().contains(leaf) {
+                    return Ok(None);
+                }
+                fetch_block_by_hash(db, tip)
```

Comment:
This function will not do what it's supposed to do. 
What this does is return the tip block, if the utxo was spent.

You need to do the following:

1.  **Get the UTXO and leaf index** If this exists, continue, else exit
2.  **Get the tip and see if it's deleted** If it is, continue, else exit
3.  **Loopback through the block accumulated data and check when it was deleted** The last block with that information in it, will be the block the TXO was spent in

I don't think we have an exact lookup for this currently so we will have to search for this.

Reply:
Thanks @SWvheerden.

As per out chat I have reverted this method back to unimplemented for the time being since there doesn't seem to be an instance of it being used.

I think you're right in that we don't have an exact lookup for this currently.